### DISCUSSION TITLE
tutorials/lift-off-part4/05-resolving-a-mutation-successfully

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#20] [course feedback]
# tutorials/lift-off-part4/05-resolving-a-mutation-successfully

https://www.apollographql.com/tutorials/lift-off-part4/05-resolving-a-mutation-successfully

**Comment by reactivicky** [comment]
Why don't we use await for other api return calls ?
<pre><code>module: async (_, { id }, { dataSources }) => {
      return await dataSources.trackAPI.getModule(id);
    },</code></pre>

instead of 
<pre><code>module: (_, { id }, { dataSources }) => {
      return dataSources.trackAPI.getModule(id);
    },</code></pre>

**Reply by R4ph-t** [reply]
Hello @reactivicky, we didn't need to use async/await before as [Apollo server can handle Promises as resolvers return type](https://www.apollographql.com/docs/apollo-server/data/resolvers#return-values). 
 `datasources.trackAPI.getModule(id)` returns a promise, and because we don't need to manipulate the result data itself in this case we can just return the promise directly. 
However for the mutation covered in this lesson, we need to know the actual result of the API call before we can return anything meaningful as the `IncrementTrackViewsResponse` type expected by our schema, hence the async/await.

**Comment by mathoks** [comment]
How do we resolve a mutation that returns an abstract interface type... How do we map the object to the types implementing the interface

**Reply by mabuyo** [reply]
Hey there @mathoks - apologies for the delay. We don't cover that topic in this particular course, however we have a course on [Intermediate Schema Design](https://www.apollographql.com/tutorials/side-quest-intermediate-schema-design/04-interfaces), with a lesson all about interfaces. Hope that helps!

**Comment by asigithub** [comment]
Why do we use a try/catch block and not promises? Promise example:
```
const {mission, spaceship} = await dataSources.spaceAPI.assignSpaceshipToMission({spaceshipId, missionId})
.then()
.catch()
```

**Reply by mabuyo** [reply]
Hey @asigithub - you should be able to use a Promise as well, we just found the try/catch approach to be more readable. You can use either one though!

**Comment by neogeek13** [comment]
I got really hung up on the code quiz as the prior example was simply "track" vs. "track: value" (the code practice return was looking for         spaceship: spaceship, and mission: mission and I was missing the : value aspects). Maybe this is just my lack of familiarity with TS - spend more of my time in mobile but perhaps helpful to bridge that gap for others.

**Reply by mabuyo** [reply]
Hey @neogeek13! To clarify, both answers are valid and it's a JavaScript concept called object literal shorthand. ([See this example for reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#property_definitions))

So in the code challenge, both answers below are valid (just showing the snippet, not the full answer

```
return {
  spaceship: spaceship,
  mission: mission
}
```
Or
```
return {
  spaceship,
  mission
}
```

**Reply by neogeek13** [reply]
Hrm. There must have been some other tiny tweak I was missing (comma. paren, something...) because it kept failing that section. When comparing with final that was the thing that stood out. Copied that small chunk in and it passed. But I see just modifying to have the values in the code exercise above also works. Thanks for clarifying.

**Comment by azimaro** [comment]
In the code challenge, it is saying that my resolver is ok, but not my return function. And it is not showing me the solution so I can figure out what's wrong.
Here's my code:
const resolvers = {
  Mutation: {
    assignSpaceship: async (_, { spaceshipId, missionId }, { dataSources }) => {
      const spaceship = await dataSources.spaceAPI.assignSpaceshipToMission(spaceshipId, missionId);
      return {
        code: 200,
        success: true,
        message: `Successfully assigned spaceship ${spaceshipId} to mission ${missionId}`,
        spaceship,
        mission
      };
    },
  },
}

**Reply by lizhennessy** [reply]
Hi there @azimaro ! It looks like we've got a bug in this challenge. I'll take a look!

**Reply by lizhennessy** [reply]
Okay, I'm back - the problem is that the challenge *should* show the solution after 3 attempts, but it's currently having an issue. I'll look into that further so it doesn't cause an issue for future learners!

For now, though, I see the problem with your code: try changing the line that calls the data source method to following.

```
const { spaceship, mission } = await dataSources.spaceAPI.assignSpaceshipToMission(spaceshipId, missionId)
```

This method actually _returns_ a spaceship and a mission in one, so that we can pass them into our return object. That should take care of passing the challenge for you! Let me know if you have any other questions. :)

**Reply by azimaro** [reply]
Doh! Of course! And I was struggling with what the const variable should be, and why there were squiggly lines under both spaceship and mission inside the return {}! I should have put 2 and 2 together, lol. Thanks!

**Reply by lizhennessy** [reply]
Yay! Happy to help! :)
